crates/core/tedge_actors/src/actors.rs

```diff
@@ -0,0 +1,181 @@
+use crate::ChannelError;
+use crate::{DynSender, Mailbox, Message};
+use async_trait::async_trait;
+
+/// Enable a struct to be used as an actor.
+///
+///
+#[async_trait]
+pub trait Actor: 'static + Sized + Send + Sync {
+    /// Type of input messages this actor consumes
+    type Input: Message;
+
+    /// Type of output messages this actor produces
+    type Output: Message;
+
+    /// Type of the mailbox used by this actor
+    type Mailbox: From<Mailbox<Self::Input>> + Send + Sync;
```

Comment:
How about making the `Mailbox` an `Iterator<Input>` or even a `Stream<Input>` so that it's easier for us to implement a concrete `Mailbox` abstraction implementing `Iterator<Input>` over multiple `Receiver`s.

Reply:
My former intent was to have something along the line of `Stream<Input>` but this issue is to let a specific implementation being able to have not one but several streams of specific messages (e.g. one main stream for asynchronous messages plus another channel to await responses of a peer). This is why I proposed to add a `From` constraint. The actor doesn't need a stream of inputs but a record of streams extracted from a stream of inputs.

Reply:
And things get more complex when trying to combine `Mailbox` and `Peers` into a single abstraction.

Reply:
The mailbox abstraction has been replaced by a more general message box abstraction that encompass message reception and sending. Doing so gives a lot of expression power to a message box implementation. The most interesting example is the [`ConcurrentServiceMessageBox`](https://github.com/thin-edge/thin-edge.io/pull/1622/files#diff-85d9e5c8a7a180c4ba09343d893d980d879c01f58dd376d6414d477a9b626cefR154) that pauses returning received requests to the actor when there are too many pending requests for which a response is still awaited.